### ISSUE TITLE
Turning monitors on and off adds 10 to workspaces 

### ISSUE THREAD

**Issue by repomansez** [sopa0/hyprsome#9]
I have a two monitor setup, when I turn them off and back on again, sometimes hyprland workspace 4 will, for example, instead of taking me to workspace 4, it takes me to workspace 24. Doesn't happen using the official hyprland workspace dispatcher.

**Comment by quqkuk** [comment]
I can confirm that's the case, as the code just prepends the id of the monitor to the number of the workspace

**Comment by sopa0** [comment]
> I can confirm that's the case, as the code just prepends the id of the monitor to the number of the workspace, it never reads hyprland's config even though the README says that you need to add the workspace rules

Yeah, and doing it like that feels hacky and stupid. Would you happen to have a better recommendation ?

**Comment by quqkuk** [comment]
Tbh I don't have many ideas as I haven't read enough documentation but you can either make a plugin or you could try using https://github.com/hyprland-community/hyprparse to parse the configuration (but I feel like that may slow down things).
I can help, but not next week

**Comment by sopa0** [comment]
I think I'll use hyprland-rs' event_listeners to detect when a workspace is created and automatically assign it to the currently focused monitor. I'll then use another event_listener to detect a monitor disconnect and move all the workspaces to one of the monitors that's still plugged. This approach will make the specific hyprsome config file obsolete, and will probably allow for an infinite number of workspaces per monitor.

I'll work on that this evening and tomorrow morning, but it may not be released until next week

**Comment by sopa0** [comment]
> I have a two monitor setup, when I turn them off and back on again, sometimes hyprland workspace 4 will, for example, instead of taking me to workspace 4, it takes me to workspace 24. No fix unless restarting Hyprland (as far as I know).
> 
> Doesn't happen using the official hyprland workspace dispatcher.


I implemented something that kinda fixes the issue:

when a workspace is added, the workspace is renamed to include the monitor's name and moves the workspace to the current monitor.

when a monitor is removed, I find all workspaces that were on the removed monitor (using the workspace names that contain the monitor's name) and all clients (windows) in those workspaces. I move all these clients to the monitor that's still plugged

**Comment by quqkuk** [comment]
> when a workspace is added, the workspace is renamed to include the monitor's name and moves the workspace to the current monitor.
> 
> when a monitor is removed, I find all workspaces that were on the removed monitor (using the workspace names that contain the monitor's name) and all clients (windows) in those workspaces. I move all these clients to the monitor that's still plugged

That means that when my monitor is unplugged everything moves to my second screen, that doesn't have the same name.
Other monitors that may be plugged in when one is unplugged can't have the same name (or don't usually have the same name) because the name is generated from the port and the order in which ports are scanned, which means that all name should be unique.
What I would do, if you want to keep this idea, is to use named workspaces and just prepend the name of the monitor to the number of the workspace, then when my monitor is unplugged and plugged back into the same port, hyprsome will just use the same set of workspaces, because even if the Monitor ID has changed, the Monitor Name will have remained the same.
This way, you don't even need to listen to events

**Comment by sopa0** [comment]
> > when a workspace is added, the workspace is renamed to include the monitor's name and moves the workspace to the current monitor.
> > when a monitor is removed, I find all workspaces that were on the removed monitor (using the workspace names that contain the monitor's name) and all clients (windows) in those workspaces. I move all these clients to the monitor that's still plugged
> 
> That means that when my monitor is unplugged everything will be moved to my second screen, that doesn't have the same name. Other monitors that may be plugged in when one is unplugged can't have the same name (or don't usually have the same name) because the name is generated from the port and the order in which ports are scanned, which means that all names _should_ be unique. What I would do, if you want to keep this idea, is to use named workspaces and just prepend the name of the monitor to the number of the workspace, then when my monitor is unplugged and plugged back into the same port, hyprsome will just use the same set of workspaces, because even if the Monitor ID has changed, the Monitor Name will have remained the same. This way, you don't even need to listen to events

I tried doing that, but using worspace names only makes the ids all fucked up 
try running 

```
hyprctl dispatch workspace name:foo
hyprctl workspaces
```

**Comment by repomansez** [comment]
> > I have a two monitor setup, when I turn them off and back on again, sometimes hyprland workspace 4 will, for example, instead of taking me to workspace 4, it takes me to workspace 24. No fix unless restarting Hyprland (as far as I know).
> > Doesn't happen using the official hyprland workspace dispatcher.
> 
> I implemented something that kinda fixes the issue:
> 
> when a workspace is added, the workspace is renamed to include the monitor's name and moves the workspace to the current monitor.
> 
> when a monitor is removed, I find all workspaces that were on the removed monitor (using the workspace names that contain the monitor's name) and all clients (windows) in those workspaces. I move all these clients to the monitor that's still plugged

what I've seen other projects do is instead of moving workspaces to the monitor that's still plugged, move the windows that were in those workspaces to the same workspaces in the plugged monitor

**Comment by sopa0** [comment]
> what I've seen other projects do is instead of moving workspaces to the monitor that's still plugged, move the windows that were in those workspaces to the same workspaces in the plugged monitor

Yeah so basically, with named workspaces, you have

eDP-1:1, eDP-1:2, etc..
HDMI-A-1:1, HDMI-A-1:2, etc...

if I unplug HDMI-A-1, I move all clients from HDMI-A-1:1 to eDP-1:1, from HDMI-A-1:2 to eDP-1:2 etc. 

I won't be able to work on that until next tuesday as I don't have access to my pc rn, but I'll be able to push a branch implementing that for testing
